numba/np/arrayobj.py

```diff
@@ -1774,6 +1774,63 @@ def imp(ary):
     return res
 
 
+def _alloc_out(a, X):
+    # dummy function that takes the arg we need to resolved (X)
+    # and the information needed to do so (a)
+    pass
+
+
+@overload(_alloc_out)
+def _alloc_out_impl(a, X):
+    # overload the dummy function to return empty_like(a) if X is None else X
+    # with type-specific implementations, in order to fully determine type of X
+    # https://github.com/numba/numba/pull/3468#issuecomment-437974147
+    if X is None or isinstance(X, types.NoneType):
+        def impl(a, X):
+            return np.empty_like(a)
+    else:
+        def impl(a, X):
+            return X
+    return impl
+
+
+@overload(np.clip)
+def np_clip(a, a_min, a_max, out=None):
```

Comment:
The types of the input arguments need validation and then tests adding to check.

Reply:
Added.